### DISCUSSION TITLE
Building locally and then pushing to remote host

### DISCUSSION THREAD

**Discussion by martynas-developer** [deployphp/deployer#2828] [Help needed] [answered]
Im trying to setup deployer 7.* but ran into a problem.

I want to build locally and then deploy the files to a remote host.
On version 6 you could use defined your task like this:
```
task('build', function () {
    invoke('deploy:prepare');
})->local();
```
But version 7 has removed local() function, and i can't figure out a way to invoke tasks locally.

I want to setup a similar CI pipeline as mentioned in this article https://dev.to/jszutkowski/how-to-automate-deploys-with-gitlab-ci-cd-and-deployer-28pb but on version 7.

**Comment by russellfeeed** [comment, marked as answer]
Hi

this page https://deployer.org/docs/7.x/UPGRADE  suggests

> Replace local() tasks with combination of once() and runLocally() func.

**Reply by martynas-developer** [reply]
yea I don't quite understand how that is supposed to fix it. I know runLocally() works with simple commands, but if I want to execute a task with many steps, it doesn't. 

I'm sure there is a way to solve this issue.

**Reply by antonmedv** [reply]
Yes, define a new localhost() and run this task on it.

**Reply by antonmedv** [reply]
Before local() would just create it behind the scene. Which leads to bunch of errors and unexpected behaviour.

**Reply by martynas-developer** [reply]
> Yes, define a new localhost() and run this task on it.

That sounds promising. 
But can you run one task on localhost() and another on remote host()? If so can you please provide an example?

**Reply by antonmedv** [reply]
Yes, you can. Will write an example.

**Reply by martynas-developer** [reply]
Thank you, that would help a lot!

**Reply by SimJoSt** [reply]
@antonmedv where did you publish the example? If you haven't done so yet, it would be greatly appreciated, if you could still do that with your current recommendation on how to do that.

**Reply by SimJoSt** [reply]
We just found an interesting sentence in the `rsync` contrib recipe: https://deployer.org/docs/7.x/contrib/rsync#rsyncwarmup:~:text=Set%20rsync_src%20to%20locally%20cloned%20repository%20and%20rsync%20to%20rsync_dest.%20Then%20set%20this%20task%20instead%20of%20deploy%3Aupdate_code%20in%20Your%20deploy%20task%20if%20Your%20hosting%20provider%20does%20not%20allow%20git.
It feels like that should be part of an example in the main part of the docs.

**Comment by brotkrueml** [comment]
I have it done this way. `ci` is localhost, `testing` and `production` are the remote hosts.

    task('build', [
        // The tasks to be executed locally
    ])->select('ci');

    task('release', [
        // The tasks to be executed remote
    ])->select('testing,production');

    task('deploy', [
        'build',
        'release',
        'deploy:cleanup',
        'deploy:success',
   ]);

Then running this command for deploying on testing:

    bin/dep deploy ci,testing --tag=v1.2.3

Or on production:

    bin/dep deploy ci,production --tag=v1.2.3

**Reply by SimJoSt** [reply]
Thank you for this snippet. It helped us out a lot to figure out how some of the concepts work.
